packages/apollo-link-dedup/src/dedupLink.ts

```diff
@@ -42,36 +45,38 @@ export class DedupLink extends ApolloLink {
       const sharedObserver = new Observable(observer => {
         // this will still be called by each subscriber regardless of
         // deduplication status
-        let prev = this.subscribers.get(key);
-        if (!prev) prev = { next: [], error: [], complete: [] };
-
-        this.subscribers.set(key, {
-          next: prev.next.concat([observer.next.bind(observer)]),
-          error: prev.error.concat([observer.error.bind(observer)]),
-          complete: prev.complete.concat([observer.complete.bind(observer)]),
-        });
+        const prevs = this.subscribers.get(key) || new Set();
+        this.subscribers.set(key, prevs.add(observer));
 
         if (!subscription) {
           subscription = singleObserver.subscribe({
             next: result => {
               const previous = cleanup(key);
-              this.subscribers.delete(key);
               if (previous) {
-                previous.next.forEach(next => next(result));
-                previous.complete.forEach(complete => complete());
+                previous.forEach(prev => {
+                  prev.next(result);
+                  prev.complete();
+                });
               }
             },
             error: error => {
               const previous = cleanup(key);
-              this.subscribers.delete(key);
-              if (previous) previous.error.forEach(err => err(error));
+              if (previous) {
+                previous.forEach(prev => prev.error(error));
+              }
             },
           });
         }
 
         return () => {
-          if (subscription) subscription.unsubscribe();
-          this.inFlightRequestObservables.delete(key);
+          const observers = this.subscribers.get(key);
+          observers.delete(observer);
+          if (observers.size === 0) {
```

Comment:
I think the length of the `observer` set and the number of times the `unsubscribe callback` is called should always be the same.So there will be no case where deleting the `non-existent key` causes the set to be empty.And the operation in the **else branch** depends on `observers.delete(observer)` Side effects, if put it in the **if condition**,it will be implicit and not semantic